src/adapters/a4g.js

```diff
@@ -0,0 +1,177 @@
+  let currentWindow = window,
+    iframeNestingLevel = 0,
+    hasExternalMeet = false,
+    hostHref = window.location.href;
+
+  while (isInIframe(currentWindow)) {
+    currentWindow = currentWindow.parent;
+
+    try {
+      if (hasExternalMeet) {
+        iframeNestingLevel = 1;
+      } else {
+        iframeNestingLevel++;
+      }
+
+      hostHref = currentWindow.document.referrer || currentWindow.location.href;
+    } catch (e) {
+      hasExternalMeet = true;
+    }
+  }
+
+  return {
+    nestingLevel: iframeNestingLevel,
+    hostHref: hostHref
+  };
+}
+
+function generateJsonpCallbackName() {
+  return '__$$A4G' + Math.random();
+}
+
+function isValidStatus(status) {
+  return status === 200;
+}
+
+function jsonp(url, callback) {
+  const scriptEl = document.createElement('SCRIPT'),
+    callbackName = generateJsonpCallbackName();
+
+  scriptEl.type = 'text/javascript';
+  scriptEl.async = true;
+  scriptEl.src = appendUrlParam(url, JSONP_PARAM_NAME, callbackName);
+
+  window[callbackName] = ({ status, response }) => {
+    isValidStatus(status)
+      ? callback(new Error(`Failed fetching ad with status ${status}`), response)
+      : callback(null, response);
+    delete window[callbackName];
+  };
+}
+
+function bidParamsToQuery(params) {
+  return A4G_SUPPORTED_PARAMS
+    .reduce((url, paramName) => paramName in params
+      ? appendUrlParam(url, paramName, params[paramName])
+      : url,
+    '');
+}
+
+function buildBidRequestUrl(bidDeliveryUrl, params) {
+  return bidDeliveryUrl + bidParamsToQuery(params);
+}
+
+function createBidRequest(bidRequest) {
+  return (status) => bidfactory.createBid(status, bidRequest);
+}
+
+function mapBidToPrebidFormat(bidRequest, bid) {
+  const bidResponse = bidRequest(constants.STATUS.GOOD);
+
+  bidResponse.bidderCode = A4G_BIDDER_CODE;
+  bidResponse.cpm = bid.cpm;
+  bidResponse.ad = bid.ad;
+  bidResponse.width = bid.width;
+  bidResponse.height = bid.height;
+
+  return bidResponse;
+}
+
+function mapBidErrorToPrebid(bidRequest) {
+  return bidRequest(constants.STATUS.NO_BID);
+}
+
+function extractBidParams(bids) {
+  const idParams = [];
+  const sizeParams = [;
+  const zoneIds = [];
+
+  let deliveryUrl = '';
+
+  for (let i = 0; i < bids.length; i++) {
+    const bid = bids;
+    if (!deliveryUrl && typeof bid.params.deliveryUrl === 'string') {
+      deliveryUrl = bid.params.deliveryUrl;
+    }
+    idParams.push(bid.adUnitCode);
+    sizeParams.push(bid.sizes.map(size => size.join(SIZE_SEPARATOR)).join(ARRAY_SIZE_SEPARATOR));
+    zoneIds.push(bid.params.zoneId);
+  }
+
+  return [deliveryUrl, {
+    [ID_PARAM_NAME]: idParams.join(ARRAY_PARAM_SEPARATOR),
+    [ZONE_ID_PARAM_NAME]: zoneIds.join(ARRAY_PARAM_SEPARATOR),
+    [SIZE_PARAM_NAME]: sizeParams.join(ARRAY_PARAM_SEPARATOR)
+  }];
+}
+
+function a4gBidFactory() {
+
+  let idCounter = 0;
+
+  function generateId() {
+    return ++idCounter;
+  }
+
+  return {
+    callBids({ params }) {
```

Comment:
Excuse me for a js noob question: 
I got used to see object literals that look like `{methodName: function() {} }` but this one look a bit different, wonder how it works